docs/src/content/docs/configuration.md
---
title: Configuration
description: Configure Embucket.
---

Embucket can be configured using environment variables or a configuration file (`.env`). Configuration file is loaded from the current directory and consist of key-value pairs. Environment variables take precedence over configuration file.

```bash
cat .env.example
# Shared settings (if no env vars are used, it will default to the first option (memory) with the shared settings)
# Iceberg Catalog settings
# Set to your catalog url
CATALOG_URL=http://127.0.0.1:3000
# Optional: CORS settings
CORS_ENABLED=true
CORS_ALLOW_ORIGIN=http://127.0.0.1:8080

# Option 1 (Memory)
# SlateDB storage settings
OBJECT_STORE_BACKEND=memory

# Option 2 (File)
# SlateDB storage settings
#OBJECT_STORE_BACKEND=file
#FILE_STORAGE_PATH=storage
#SLATEDB_PREFIX=state

# Option 3 (S3)
# SlateDB storage settings
#OBJECT_STORE_BACKEND=s3
# Optional: AWS S3 storage (leave blank if using local storage)
#AWS_ACCESS_KEY_ID="<your_aws_access_key_id>"
#AWS_SECRET_ACCESS_KEY="<your_aws_secret_access_key>"
#AWS_REGION="<your_aws_region>"
#S3_BUCKET="<your_s3_bucket>"
#S3_ALLOW_HTTP=
```

```bash
$ embucketd --help
Usage: embucketd [OPTIONS]

Options:
  -b, --backend <BACKEND>
          Backend to use for state storage [env: OBJECT_STORE_BACKEND=file] [default: memory] [possible values: s3, file, memory]
  -s, --slatedb-prefix <SLATEDB_PREFIX>
          [env: SLATEDB_PREFIX=data/] [default: state]
      --host <HOST>
          Host to bind to [env: BUCKET_HOST=] [default: localhost]
      --port <PORT>
          Port to bind to [env: BUCKET_PORT=] [default: 3000]
      --assets-port <ASSETS_PORT>
          Port of web assets server to bind to [env: WEB_ASSETS_PORT=] [default: 8080]
      --catalog-url <CATALOG_URL>
          Iceberg catalog url [env: CATALOG_URL=] [default: http://127.0.0.1:3000]
      --cors-enabled <CORS_ENABLED>
          Enable CORS [env: CORS_ENABLED=true] [default: true] [possible values: true, false]
      --cors-allow-origin <CORS_ALLOW_ORIGIN>
          CORS Allow Origin [env: CORS_ALLOW_ORIGIN=http://localhost:8080] [default: http://localhost:8080]
  -d, --data-format <DATA_FORMAT>
          Data serialization format in Snowflake v1 API [env: DATA_FORMAT=json] [default: json]
      --jwt-secret <JWT_SECRET>
          JWT secret for auth [env: JWT_SECRET]
      --auth-demo-user <AUTH_DEMO_USER>
          User for auth demo [env: AUTH_DEMO_USER=] [default: embucket]
      --auth-demo-password <AUTH_DEMO_PASSWORD>
          Password for auth demo [env: AUTH_DEMO_PASSWORD=] [default: embucket]
      --tracing-level <TRACING_LEVEL>
          Tracing level, it can be overrided by *RUST_LOG* env var [env: TRACING_LEVEL=] [default: info] [possible values: off, info, debug, trace]
      --tracing-span-processor <TRACING_SPAN_PROCESSOR>
          Tracing span processor [env: span_processor=] [default: batch-span-processor] [possible values: batch-span-processor, batch-span-processor-experimental-async-runtime]
  -h, --help
          Print help
  -V, --version
          Print version

S3 Backend Options:
      --access-key-id <ACCESS_KEY_ID>
          AWS Access Key ID [env: AWS_ACCESS_KEY_ID=]
      --secret-access-key <SECRET_ACCESS_KEY>
          AWS Secret Access Key [env: AWS_SECRET_ACCESS_KEY=]
      --region <REGION>
          AWS Region [env: AWS_REGION=]
      --bucket <BUCKET>
          S3 Bucket Name [env: S3_BUCKET=]
      --endpoint <ENDPOINT>
          S3 Endpoint (Optional) [env: S3_ENDPOINT=]
      --allow-http <ALLOW_HTTP>
          Allow HTTP for S3 (Optional) [env: S3_ALLOW_HTTP=] [possible values: true, false]

File Backend Options:
      --file-storage-path <FILE_STORAGE_PATH>
          Path to the directory where files will be stored [env: FILE_STORAGE_PATH=data/]
```

## Generic options

`--no-bootstrap` option is used to disable bootstrap process. Bootstrap routine creates in-memory volume, default database (named `embucket`) and default schema (named `public`).

`DATA_FORMAT` option is used to set data format in Snowflake v1 API. It can be `json` or `arrow`.

## HTTP configuration

Embucket uses HTTP to serve the UI and API. This is configured using `HOST` and `PORT` environment variables. Once launched, API is available at `http://HOST:PORT`.

## Metadata configuration

Embucket uses SlateDB to persist metadata. This is configured using `SLATEDB_PREFIX` and `OBJECT_STORE_BACKEND` environment variables.
There are three options for SlateDB storage:

1. Memory
2. File system
3. Object storage

Memory and file system backends are for testing and development purposes. Object storage is recommended for production use.

When using file as a backend, `FILE_STORAGE_PATH` environment variable should be set to the directory where files will be stored. `SLATEDB_PREFIX` environment variable should be set to the prefix of the SlateDB database.

When using object storage, `OBJECT_STORE_BACKEND` environment variable should be set to `s3`. `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `S3_BUCKET`, `S3_ENDPOINT`, `S3_ALLOW_HTTP` environment variables should be set to the values of the object storage provider.

## UI configuration

Embucket uses `HOST` and `ASSETS_PORT` environment variables to serve the UI and API. Once launched, you can access the UI at `http://HOST:ASSETS_PORT`.
`AUTH_DEMO_USER` and `AUTH_DEMO_PASSWORD` environment variables are used to authenticate the demo user (credentials to login in the UI).

User is required to set `JWT_SECRET` environment variable to enable authentication (this should be set to some random string).

`CORS_ENABLED` and `CORS_ALLOW_ORIGIN` environment variables are used to configure CORS. `CORS_ENABLED` should be set to `true` to enable CORS. `CORS_ALLOW_ORIGIN` should be set to the origin of the UI (e.g. `http://localhost:8080` or `https://acme.com`).

## Iceberg catalog configuration

Embucket provides an Iceberg catalog REST API at `http://HOST:PORT/catalog`. `CATALOG_URL` environment variable should be set to the URL of the Iceberg catalog, this is required for external clients to be able to connect to the catalog when requesting connection options via `v1/config` endpoint.


docs/planetary-halo/src/content/docs/configuration.md
---
title: Configuration
description: Configure Embucket.
---

Embucket can be configured using environment variables or a configuration file (`.env`). Configuration file is loaded from the current directory and consist of key-value pairs. Environment variables take precedence over configuration file.

```bash
cat .env.example
# Shared settings (if no env vars are used, it will default to the first option (memory) with the shared settings)
# Iceberg Catalog settings
# Set to your catalog url
CATALOG_URL=http://127.0.0.1:3000
# Optional: CORS settings
CORS_ENABLED=true
CORS_ALLOW_ORIGIN=http://127.0.0.1:8080

# Option 1 (Memory)
# SlateDB storage settings
OBJECT_STORE_BACKEND=memory

# Option 2 (File)
# SlateDB storage settings
#OBJECT_STORE_BACKEND=file
#FILE_STORAGE_PATH=storage
#SLATEDB_PREFIX=state

# Option 3 (S3)
# SlateDB storage settings
#OBJECT_STORE_BACKEND=s3
# Optional: AWS S3 storage (leave blank if using local storage)
#AWS_ACCESS_KEY_ID="<your_aws_access_key_id>"
#AWS_SECRET_ACCESS_KEY="<your_aws_secret_access_key>"
#AWS_REGION="<your_aws_region>"
#S3_BUCKET="<your_s3_bucket>"
#S3_ALLOW_HTTP=
```

```bash
$ embucketd --help
Usage: embucketd [OPTIONS]

Options:
  -b, --backend <BACKEND>
          Backend to use for state storage [env: OBJECT_STORE_BACKEND=file] [default: memory] [possible values: s3, file, memory]
  -s, --slatedb-prefix <SLATEDB_PREFIX>
          [env: SLATEDB_PREFIX=data/] [default: state]
      --host <HOST>
          Host to bind to [env: BUCKET_HOST=] [default: localhost]
      --port <PORT>
          Port to bind to [env: BUCKET_PORT=] [default: 3000]
      --assets-port <ASSETS_PORT>
          Port of web assets server to bind to [env: WEB_ASSETS_PORT=] [default: 8080]
      --catalog-url <CATALOG_URL>
          Iceberg catalog url [env: CATALOG_URL=] [default: http://127.0.0.1:3000]
      --cors-enabled <CORS_ENABLED>
          Enable CORS [env: CORS_ENABLED=true] [default: true] [possible values: true, false]
      --cors-allow-origin <CORS_ALLOW_ORIGIN>
          CORS Allow Origin [env: CORS_ALLOW_ORIGIN=http://localhost:8080] [default: http://localhost:8080]
  -d, --data-format <DATA_FORMAT>
          Data serialization format in Snowflake v1 API [env: DATA_FORMAT=json] [default: json]
      --jwt-secret <JWT_SECRET>
          JWT secret for auth [env: JWT_SECRET]
      --auth-demo-user <AUTH_DEMO_USER>
          User for auth demo [env: AUTH_DEMO_USER=] [default: embucket]
      --auth-demo-password <AUTH_DEMO_PASSWORD>
          Password for auth demo [env: AUTH_DEMO_PASSWORD=] [default: embucket]
      --tracing-level <TRACING_LEVEL>
          Tracing level, it can be overrided by *RUST_LOG* env var [env: TRACING_LEVEL=] [default: info] [possible values: off, info, debug, trace]
      --tracing-span-processor <TRACING_SPAN_PROCESSOR>
          Tracing span processor [env: span_processor=] [default: batch-span-processor] [possible values: batch-span-processor, batch-span-processor-experimental-async-runtime]
  -h, --help
          Print help
  -V, --version
          Print version

S3 Backend Options:
      --access-key-id <ACCESS_KEY_ID>
          AWS Access Key ID [env: AWS_ACCESS_KEY_ID=]
      --secret-access-key <SECRET_ACCESS_KEY>
          AWS Secret Access Key [env: AWS_SECRET_ACCESS_KEY=]
      --region <REGION>
          AWS Region [env: AWS_REGION=]
      --bucket <BUCKET>
          S3 Bucket Name [env: S3_BUCKET=]
      --endpoint <ENDPOINT>
          S3 Endpoint (Optional) [env: S3_ENDPOINT=]
      --allow-http <ALLOW_HTTP>
          Allow HTTP for S3 (Optional) [env: S3_ALLOW_HTTP=] [possible values: true, false]

File Backend Options:
      --file-storage-path <FILE_STORAGE_PATH>
          Path to the directory where files will be stored [env: FILE_STORAGE_PATH=data/]
```

## Generic options

`--no-bootstrap` option is used to disable bootstrap process. Bootstrap routine creates in-memory volume, default database (named `embucket`) and default schema (named `public`).

`DATA_FORMAT` option is used to set data format in Snowflake v1 API. It can be `json` or `arrow`.

## HTTP configuration

Embucket uses HTTP to serve the UI and API. This is configured using `HOST` and `PORT` environment variables. Once launched, API is available at `http://HOST:PORT`.

## Metadata configuration

Embucket uses SlateDB to persist metadata. This is configured using `SLATEDB_PREFIX` and `OBJECT_STORE_BACKEND` environment variables.
There are three options for SlateDB storage:

1. Memory
2. File system
3. Object storage

Memory and file system backends are for testing and development purposes. Object storage is recommended for production use.

When using file as a backend, `FILE_STORAGE_PATH` environment variable should be set to the directory where files will be stored. `SLATEDB_PREFIX` environment variable should be set to the prefix of the SlateDB database.

When using object storage, `OBJECT_STORE_BACKEND` environment variable should be set to `s3`. `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `S3_BUCKET`, `S3_ENDPOINT`, `S3_ALLOW_HTTP` environment variables should be set to the values of the object storage provider.

## UI configuration

Embucket uses `HOST` and `ASSETS_PORT` environment variables to serve the UI and API. Once launched, you can access the UI at `http://HOST:ASSETS_PORT`. 
`AUTH_DEMO_USER` and `AUTH_DEMO_PASSWORD` environment variables are used to authenticate the demo user (credentials to login in the UI).

User is required to set `JWT_SECRET` environment variable to enable authentication (this should be set to some random string).

`CORS_ENABLED` and `CORS_ALLOW_ORIGIN` environment variables are used to configure CORS. `CORS_ENABLED` should be set to `true` to enable CORS. `CORS_ALLOW_ORIGIN` should be set to the origin of the UI (e.g. `http://localhost:8080` or `https://acme.com`).


## Iceberg catalog configuration

Embucket provides an Iceberg catalog REST API at `http://HOST:PORT/catalog`. `CATALOG_URL` environment variable should be set to the URL of the Iceberg catalog, this is required for external clients to be able to connect to the catalog when requesting connection options via `v1/config` endpoint.

crates/embucketd/src/cli.rs
use clap::{Parser, ValueEnum};
use object_store::{
    ObjectStore, Result as ObjectStoreResult, aws::AmazonS3Builder, aws::S3ConditionalPut,
    local::LocalFileSystem, memory::InMemory,
};
use std::fs;
use std::path::PathBuf;
use std::sync::Arc;
use tracing_subscriber::filter::LevelFilter;

#[derive(Parser)]
#[command(version, about, long_about=None)]
pub struct CliOpts {
    #[arg(
        long,
        env = "NO_BOOTSTRAP",
        default_value = "false",
        help = "Disable bootstrap functionality"
    )]
    pub no_bootstrap: bool,

    #[arg(
        short,
        long,
        value_enum,
        env = "OBJECT_STORE_BACKEND",
        default_value = "memory",
        help = "Backend to use for state storage"
    )]
    backend: StoreBackend,

    #[arg(
        long,
        env = "AWS_ACCESS_KEY_ID",
        required_if_eq("backend", "s3"),
        help = "AWS Access Key ID",
        help_heading = "S3 Backend Options"
    )]
    access_key_id: Option<String>,
    #[arg(
        long,
        env = "AWS_SECRET_ACCESS_KEY",
        required_if_eq("backend", "s3"),
        help = "AWS Secret Access Key",
        help_heading = "S3 Backend Options"
    )]
    secret_access_key: Option<String>,
    #[arg(
        long,
        env = "AWS_REGION",
        required_if_eq("backend", "s3"),
        help = "AWS Region",
        help_heading = "S3 Backend Options"
    )]
    region: Option<String>,
    #[arg(
        long,
        env = "S3_BUCKET",
        required_if_eq("backend", "s3"),
        help = "S3 Bucket Name",
        help_heading = "S3 Backend Options"
    )]
    bucket: Option<String>,
    #[arg(
        long,
        env = "S3_ENDPOINT",
        help = "S3 Endpoint (Optional)",
        help_heading = "S3 Backend Options"
    )]
    endpoint: Option<String>,
    #[arg(
        long,
        env = "S3_ALLOW_HTTP",
        help = "Allow HTTP for S3 (Optional)",
        help_heading = "S3 Backend Options"
    )]
    allow_http: Option<bool>,

    #[arg(
        long,
        env = "FILE_STORAGE_PATH",
        required_if_eq("backend", "file"),
        help_heading = "File Backend Options",
        help = "Path to the directory where files will be stored"
    )]
    file_storage_path: Option<PathBuf>,

    #[arg(short, long, env = "SLATEDB_PREFIX", default_value = "state")]
    pub slatedb_prefix: String,

    #[arg(
        long,
        env = "BUCKET_HOST",
        default_value = "localhost",
        help = "Host to bind to"
    )]
    pub host: Option<String>,

    #[arg(
        long,
        env = "BUCKET_PORT",
        default_value = "3000",
        help = "Port to bind to"
    )]
    pub port: Option<u16>,

    #[arg(
        long,
        env = "WEB_ASSETS_PORT",
        default_value = "8080",
        help = "Port of web assets server to bind to"
    )]
    pub assets_port: Option<u16>,

    #[arg(
        long,
        env = "CATALOG_URL",
        default_value = "http://127.0.0.1:3000",
        help = "Iceberg catalog url"
    )]
    pub catalog_url: Option<String>,

    #[arg(
        long,
        env = "CORS_ENABLED",
        help = "Enable CORS",
        default_value = "true"
    )]
    pub cors_enabled: Option<bool>,

    #[arg(
        long,
        env = "CORS_ALLOW_ORIGIN",
        default_value = "http://localhost:8080",
        required_if_eq("cors_enabled", "true"),
        help = "CORS Allow Origin"
    )]
    pub cors_allow_origin: Option<String>,

    #[arg(
        short,
        long,
        default_value = "json",
        env = "DATA_FORMAT",
        help = "Data serialization format in Snowflake v1 API"
    )]
    pub data_format: Option<String>,

    // should unset JWT_SECRET env var after loading
    #[arg(
        long,
        env = "JWT_SECRET",
        hide_env_values = true,
        help = "JWT secret for auth"
    )]
    jwt_secret: Option<String>,

    #[arg(
        long,
        env = "AUTH_DEMO_USER",
        value_parser = clap::builder::NonEmptyStringValueParser::new(),
        default_value = "embucket",
        help = "User for auth demo"
    )]
    pub auth_demo_user: Option<String>,

    #[arg(
        long,
        env = "AUTH_DEMO_PASSWORD",
        value_parser = clap::builder::NonEmptyStringValueParser::new(),
        default_value = "embucket",
        help = "Password for auth demo"
    )]
    pub auth_demo_password: Option<String>,

    #[arg(
        long,
        value_enum,
        env = "TRACING_LEVEL",
        default_value = "info",
        help = "Tracing level, it can be overrided by *RUST_LOG* env var"
    )]
    pub tracing_level: TracingLevel,

    #[arg(
        long,
        value_enum,
        env = "span_processor",
        default_value = "batch-span-processor",
        help = "Tracing span processor"
    )]
    pub tracing_span_processor: TracingSpanProcessor,
}

#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ValueEnum)]
enum StoreBackend {
    S3,
    File,
    Memory,
}

impl CliOpts {
    #[allow(clippy::unwrap_used, clippy::as_conversions)]
    pub fn object_store_backend(self) -> ObjectStoreResult<Arc<dyn ObjectStore>> {
        match self.backend {
            StoreBackend::S3 => {
                let s3_allow_http = self.allow_http.unwrap_or(false);

                let s3_builder = AmazonS3Builder::new()
                    .with_access_key_id(self.access_key_id.unwrap())
                    .with_secret_access_key(self.secret_access_key.unwrap())
                    .with_region(self.region.unwrap())
                    .with_bucket_name(self.bucket.unwrap())
                    .with_conditional_put(S3ConditionalPut::ETagMatch);

                if let Some(endpoint) = self.endpoint {
                    s3_builder
                        .with_endpoint(&endpoint)
                        .with_allow_http(s3_allow_http)
                        .build()
                        .map(|s3| Arc::new(s3) as Arc<dyn ObjectStore>)
                } else {
                    s3_builder
                        .build()
                        .map(|s3| Arc::new(s3) as Arc<dyn ObjectStore>)
                }
            }
            StoreBackend::File => {
                let file_storage_path = self.file_storage_path.unwrap();
                let path = file_storage_path.as_path();
                if !path.exists() || !path.is_dir() {
                    fs::create_dir(path).unwrap();
                }
                LocalFileSystem::new_with_prefix(file_storage_path)
                    .map(|fs| Arc::new(fs) as Arc<dyn ObjectStore>)
            }
            StoreBackend::Memory => Ok(Arc::new(InMemory::new()) as Arc<dyn ObjectStore>),
        }
    }

    // method resets a secret env
    pub fn jwt_secret(&self) -> String {
        unsafe {
            std::env::remove_var("JWT_SECRET");
        }
        self.jwt_secret.clone().unwrap_or_default()
    }
}

#[derive(Debug, Clone, ValueEnum)]
pub enum TracingLevel {
    Off,
    Info,
    Debug,
    Trace,
}

#[derive(Debug, Clone, ValueEnum)]
pub enum TracingSpanProcessor {
    BatchSpanProcessor,
    BatchSpanProcessorExperimentalAsyncRuntime,
}

#[allow(clippy::from_over_into)]
impl Into<LevelFilter> for TracingLevel {
    fn into(self) -> LevelFilter {
        match self {
            Self::Off => LevelFilter::OFF,
            Self::Info => LevelFilter::INFO,
            Self::Debug => LevelFilter::DEBUG,
            Self::Trace => LevelFilter::TRACE,
        }
    }
}

impl std::fmt::Display for TracingLevel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Off => write!(f, "off"),
            Self::Info => write!(f, "info"),
            Self::Debug => write!(f, "debug"),
            Self::Trace => write!(f, "trace"),
        }
    }
}
